### DISCUSSION TITLE
Updating edge style dynamically

### DISCUSSION THREAD

**Discussion by bigbluemonster** [xyflow/xyflow#2406] [Q&A] [answered]
Hello all,

I am trying to update my edges dynamically but I've hit a bit of a stumbling block. I have created a custom node for mapping tables and fields, and I want to click a field and update the edges that match it's lineage to a different colour so it's easy to see.

This is what I have:

`import { useEffect, useState } from 'react';
import { Handle, Position, updateEdge } from 'react-flow-renderer';
import { contentStyle as style } from "./Node";

function FieldNode({ data, field }) {
    const [highlighted, setHighlighted] = useState(false)

    const fieldStyle = {
        highlightField: {
            border: highlighted ? "2px solid #14b8a6" : ""
        }
    }

    function useTesterFunc() {
        setHighlighted(!highlighted);
        useEffect(() => {
            updateEdge((edge) =>
                edge.map((edg) => {
                    if (edg.includes(data.id + "." + field.name)) {
                        edg.style = {
                            ...edg.syle, stroke: 'red',
                        };
                    }
                    return edg;
                })
            );
        }, [updateEdge])
    }

    return (
        <div key={data.id + "." + field.name} style={{ ...style.io, ...style.textCenter, ...fieldStyle.highlightField }}>
            <button className='w-full p-2 mb-2' id={field.name} onClick={useTesterFunc}>
                {field.name}
            </button>
            {data.type === "branchNode" || data.type === "leafNode"
                ? <Handle type="target"
                    id={"i-" + data.table + "." + field.name}
                    position={Position.Left}
                    style={{ ...style.handle, ...style.left }}
                />
                : <></>
            }
            {data.type === "rootNode" || data.type === "branchNode"
                ? <Handle type="source"
                    id={"o-" + data.table + "." + field.name}
                    position={Position.Right}
                    style={{ ...style.handle, ...style.right }}
                />
                : <></>
            }
        </div>
    );
};

export default FieldNode;`

I get this error message though after I perform the button onClick:

`react-dom.development.js:16227 -  Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
    at Object.throwInvalidHookError (react-dom.development.js:16227:1)
    at useEffect (react.development.js:1634:1)
    at useTesterFunc (FieldNode.js:25:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:4164:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4213:1)
    at invokeGuardedCallback (react-dom.development.js:4277:1)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4291:1)
    at executeDispatch (react-dom.development.js:9041:1)
    at processDispatchQueueItemsInOrder (react-dom.development.js:9073:1)
    at processDispatchQueue (react-dom.development.js:9086:1)`

I may not be understanding the documentation right on how to update the edge. Any pointers would be great!

**Comment by bcakmakoglu** [comment, marked as answer]
You're incorrectly using `useEffect`.
You can't call `useEffect` inside a function (as the error suggests), only inside the "body" of a function *component*.
Meaning: Your effects must be called inside the components "setup" instructions, not in callbacks that are triggered on user interaction.

For example: 

```jsx
const SomeComponent = ({ someProp }) => {
	/**
	 * this is the "setup" of your component
	 * here you can instruct your component to do some effects etc.
	 */
	
	// this is a valid effect
	useEffect(() => {
		// do some effect logic here
	}, [someProp]

	const onClickHandler = () => {
		/**
		* this is *NOT* inside the setup of your component anymore
		* it's inside the scope of this function so instructing your component to do an effect is not allowed
		*/

		// this is *NOT* a valid effect, it will throw an error
		useEffect(() => {
			// do some effect logic here
		}, [someProp]
	}

	return <div onClick={onClickHandler}>Foo</div>
}
```

Maybe you confused it with [`useCallback`](https://www.w3schools.com/react/react_usecallback.asp).
So you might want to change `useTesterFunc` to something like this:

```js
useCallback(() => {
  setHighlighted(!highlighted)
  // rest of your code
}, [/* effect dependencies */)
```

Furthermore `updateEdge` is meant to update the source and target of an edge, not it's styles.
Here's an [example](https://reactflow.dev/docs/examples/edges/updatable-edge/) of how `updateEdge` is meant to be used.

Instead you might want to use `setEdges`, i.e. something like this:

```js
setEdges((edges) => {
	return edges.map((edge) => {
		if (edge.id === someId) {
			return {
				...edge,
				style: { ... }, // some different styles
			}
		}
		
		return edge
	})
})

**Reply by bigbluemonster** [reply]
Thanks for taking the time to reply. It seems in my custom node file, it is throwing this error when trying to use setEdges:

`export 'setEdges' (imported as 'setEdges') was not found in 'react-flow-renderer' (possible exports: Background, BackgroundVariant, BezierEdge, ConnectionLineType, ConnectionMode, ControlButton, Controls, EdgeText, Handle, MarkerType, MiniMap, PanOnScrollMode, Position, ReactFlowProvider, SimpleBezierEdge, SmoothStepEdge, StepEdge, StraightEdge, addEdge, applyEdgeChanges, applyNodeChanges, default, getBezierEdgeCenter, getBezierPath, getConnectedEdges, getEdgeCenter, getIncomers, getMarkerEnd, getOutgoers, getRectOfNodes, getSimpleBezierEdgeCenter, getSimpleBezierPath, getSmoothStepPath, getTransformForBounds, internalsSymbol, isEdge, isNode, updateEdge, useEdges, useEdgesState, useKeyPress, useNodes, useNodesState, useReactFlow, useStore, useStoreApi, useUpdateNodeInternals, useViewport)`

Because the custom node is set by the node type defined when the nodes are loaded in, they are children of the ReactFlow component but I don't know how to access the edge state to change the edges. useEdgeState maybe?

**Reply by bcakmakoglu** [reply]
`setEdges` is not an export of `react-flow-renderer` 😅

As you noted yourself `useEdgesState` is the correct one and you'll have to find a way to access the setter from your node.
Some people resort to using stores like Redux or sorts for that but you can find your own way, this is really not a React Flow issue but something that you'll have to figure out about React itself 🤓

Check the [examples](https://reactflow.dev/docs/examples/overview/) and try to play around in the sandbox.
They should give you an idea on how to handle these things properly.

**Reply by bigbluemonster** [reply]
Ah that makes sense now! So I have to find access to the main setter from my ReactFlow set up component and once changes are made there, they will refresh. Ok - that I can figure out now I have some guidance. Thanks again!!

**Reply by bcakmakoglu** [reply]
Yes, exactly 😄 
No problem & Have fun coding!

**Reply by bigbluemonster** [reply]
Just wanted to let you know that I managed to find the answer I was looking for and now my edges turn red when my custom field node is clicked. Thanks for all the info pointing me in the right direction!

For those who stumble across this question, I used the info from [here](https://reactflow.dev/docs/guides/state-management/) and adapted it to fit my needs.

**Reply by bcakmakoglu** [reply]
@bigbluemonster if you could select the answer so this thread appears as "answered", that would be nice =) 
So others know this has been answered
